src/runtime/importhook.cs

```diff
@@ -141,189 +128,177 @@ internal static void SaveRuntimeData(RuntimeDataStorage storage)
-                        {
-                            continue;
-                        }
+                Runtime.XDecref(finder);
+                throw new PythonException();
+            }
 
-                        Runtime.XIncref(attr.pyHandle);
-                        using (var obj = new PyObject(attr.pyHandle))
-                        {
-                            mod_dict.SetItem(s, obj);
-                        }
-                    }
-                }
+            // Set as a sub-module of clr.
+            Runtime.XIncref(import_hook_module);
+            if(Runtime.PyModule_AddObject(py_clr_module, "loader", import_hook_module) != 0)
+            {
+                Runtime.XDecref(import_hook_module);
+                throw new PythonException();
             }
-            Runtime.XIncref(py_clr_module);
-            return NewReference.DangerousFromPointer(py_clr_module);
+
+            // Finally, add the hook to the meta path
+            var finder_inst = Runtime.PyDict_GetItemString(mod_dict, "finder_inst").DangerousGetAddressOrNull();
+            Runtime.XIncref(finder);
+            var metapath = Runtime.PySys_GetObject("meta_path");
+            Runtime.PyList_Append(metapath, finder);
         }
 
         /// <summary>
-        /// The actual import hook that ties Python to the managed world.
+        /// Sets up the tracking of loaded namespaces. This makes available to 
+        /// Python, as a Python object, the loaded namespaces. The set of loaded
+        /// namespaces is used during the import to verify if we can import a 
+        /// CLR assembly as a module or not. The set is stored on the clr module.
         /// </summary>
-        public static IntPtr __import__(IntPtr self, IntPtr argsRaw, IntPtr kw)
+        static void SetupNamespaceTracking()
         {
-            var args = new BorrowedReference(argsRaw);
-
-            // Replacement for the builtin __import__. The original import
-            // hook is saved as this.py_import. This version handles CLR
-            // import and defers to the normal builtin for everything else.
-
-            var num_args = Runtime.PyTuple_Size(args);
-            if (num_args < 1)
-            {
-                return Exceptions.RaiseTypeError("__import__() takes at least 1 argument (0 given)");
-            }
-
-            BorrowedReference py_mod_name = Runtime.PyTuple_GetItem(args, 0);
-            if (py_mod_name.IsNull ||
-                !Runtime.IsStringType(py_mod_name))
-            {
-                return Exceptions.RaiseTypeError("string expected");
-            }
-
-            // Check whether the import is of the form 'from x import y'.
-            // This determines whether we return the head or tail module.
-
-            BorrowedReference fromList = default;
-            var fromlist = false;
-            if (num_args >= 4)
+            var newset = Runtime.PySet_New(new BorrowedReference(IntPtr.Zero));
```

Comment:
You Dispose in a finally; should this be a `using var` ?

Reply:
NIT: I'd replace `new BorrowedReference(IntPtr.Zero)` with either `BorrowedReference.Null` or simply `default`.